iib/workers/tasks/build.py

```diff
@@ -436,6 +429,62 @@ def _opm_index_rm(base_dir, operators, binary_image, from_index, overwrite_from_
         )
 
 
+def _overwrite_from_index(
+    request_id, output_pull_spec, from_index, overwrite_from_index_token=None
+):
+    """
+    Overwrite the ``from_index`` image.
+
+    :param int request_id: the ID of the request this index image is for.
+    :param str output_pull_spec: the pull specification of the manifest list for the index image
+        that IIB built.
+    :param str from_index: the pull specification of the image to overwrite.
+    :param str overwrite_from_index_token: the user supplied token to use when overwriting the
+        ``from_index`` image. If this is not set, IIB's configured credentials will be used.
+    :raises IIBError: if one of the skopeo commands fails.
+    """
+    state_reason = f'Overwriting the index image {from_index} with {output_pull_spec}'
+    log.info(state_reason)
+    set_request_state(request_id, 'in_progress', state_reason)
+
+    new_index_src = f'docker://{output_pull_spec}'
+    temp_dir = None
+    try:
+        if overwrite_from_index_token:
+            output_pull_spec_registry = ImageName.parse(output_pull_spec).registry
+            from_index_registry = ImageName.parse(from_index).registry
+            # If the registries are the same and `overwrite_from_index_token` was supplied, that
+            # means that IIB's token will likely not have access to read the `from_index` image.
+            # This means IIB must first export the manifest list and all the manifests locally and
+            # then overwrite the `from_index` image with the exported version using the user
+            # supplied token.
+            #
+            # When a newer version of buildah is available in RHEL 8, then that can be used instead
+            # of the manifest-tool to create the manifest list locally which means this workaround
+            # can be removed.
+            if output_pull_spec_registry == from_index_registry:
+                temp_dir = tempfile.TemporaryDirectory(prefix='iib-')
+                new_index_src = f'oci:{temp_dir.name}'
+                log.info(
+                    'The registry used by IIB (%s) is also the registry where from_index (%s) will '
+                    'be overwritten using the user supplied token. Will perform a workaround which '
+                    'will cause the manifest digests to change but the content is the same.',
```

Comment:
Should this be mentioned in the state_reason so that the user is aware of the change in manifest digests?

Reply:
I see your point but the manifest digests are already going to be different since the index image was modified by either removing an operator or adding a bundle. This is logging the fact that the digests will be different between the version in IIB's configured repo and the overwritten `from_index`. It's more for administrators to be aware that something isn't broken.